### PR TITLE
Provides an extension to Crawler4J for crawling dynamic web pages

### PR DESCRIPTION
**Pull request by ngsoftwaredev**
This pull request originates from a project I did for a customer where we had to crawl one of their public websites (made with  ReactJS) for building a search engine on it. 

Starting from an initial approach in Python, it wasn't easy to maintain or to operate. So I went on taking advantage of Crawler4J since I actually hate it that everytime some dynamic website has to be crawled it's done in Python. I'd rather have a much better and easy to operate Java implementation!

It took some effort to wire together Crawler4j (which I used extensively in past projects) and a headless browser.

[The README](https://github.com/rzo1/crawler4j#reconstructing-extra-urls-to-crawl) mentions the possibility to extend the parser for such a usage scenario, so I thought it would actually be nice to make this much simpler for anyone wishing to address dynamic content crawling with Crawler4j.

The implementation makes use of Selenium's _WebDriver_ API to interface with a headless browser. It deals with the fact that _WebDriver_ is not thread-safe, so basically we instantiate one headless browser per crawler thread.

Also it offers a programmatic approach for waiting for dynamic content loading.

I'll be glad to at least have your feedback on this PR.

Things that could be done to enhance it:

- update the README (I'll definitely do it if you consider merging this)
- maybe package this feature as some sort of extension instead of having it in the core module? 

Best regards from France,

Nicolas

### CODE REVIEW

**Comment by ngsoftwaredev** [comment]
One criticism to this implementation is that it's not happening at the _PageFetcher_ level. However doing it this way would require some refactoring, while my approach is only using extension.

One bug I noticed is that I'm not enforcing `CrawlConfig#maxDownloadSize` when fetching the page with WebDriver. I should fix this.

File: crawler4j-commons/src/main/java/edu/uci/ics/crawler4j/crawler/DynamicCrawlConfig.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
+package edu.uci.ics.crawler4j.crawler;
```
Comment (rzo1) [inline crawler4j-commons/src/main/java/edu/uci/ics/crawler4j/crawler/DynamicCrawlConfig.java:1]: Misses a license header

File: crawler4j-commons/src/main/java/edu/uci/ics/crawler4j/crawler/DynamicCrawlConfig.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.util.Collections;
+import java.util.List;
+
+import static edu.uci.ics.crawler4j.crawler.DynamicCrawlConfig.WebDriverType.firefox;
```
Comment (rzo1) [inline crawler4j-commons/src/main/java/edu/uci/ics/crawler4j/crawler/DynamicCrawlConfig.java:8]: Don't think we need that static import? We can just reference the full qualifier in line 20

File: crawler4j-commons/src/main/java/edu/uci/ics/crawler4j/crawler/DynamicCrawlConfig.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+    /**
+     * If dynamic content crawling is used, the maximum time to wait for dynamic content loading in seconds.
+     */
+    private int maxWaitForDynamicContentInSeconds = 2;
+
+    /**
+     * Absolute path to the driver binary on the local filesystem.
+     */
+    private String webDriverPath;
```
Comment (rzo1) [inline crawler4j-commons/src/main/java/edu/uci/ics/crawler4j/crawler/DynamicCrawlConfig.java:30]: Can we use something like `Path` here? Less error prone than a plain string?

File: crawler4j-commons/src/main/java/edu/uci/ics/crawler4j/crawler/DynamicCrawlConfig.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+    /**
+     * The user agent to be used by the web driver.
+     */
+    private String webDriverUserAgent;
```
Comment (rzo1) [inline crawler4j-commons/src/main/java/edu/uci/ics/crawler4j/crawler/DynamicCrawlConfig.java:35]: Is there a reason for not using the value from https://github.com/rzo1/crawler4j/blob/master/crawler4j-commons/src/main/java/edu/uci/ics/crawler4j/crawler/CrawlConfig.java#L65 ? Do we have any good reasons for introducing a 2nd user agent config option?

File: crawler4j-core/src/main/java/edu/uci/ics/crawler4j/crawler/dynamic/DynamicCrawlerHelper.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
+package edu.uci.ics.crawler4j.crawler.dynamic;
```
Comment (rzo1) [inline crawler4j-core/src/main/java/edu/uci/ics/crawler4j/crawler/dynamic/DynamicCrawlerHelper.java:1]: Header missing

File: crawler4j-core/src/main/java/edu/uci/ics/crawler4j/crawler/dynamic/DynamicCrawlerHelper.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+import edu.uci.ics.crawler4j.robotstxt.RobotstxtServer;
+import edu.uci.ics.crawler4j.url.TLDList;
+import org.openqa.selenium.By;
+
+import java.util.function.BiFunction;
+
+/**
+ * Helper class that provides factory methods for a {@link CrawlController} set up for headless browser crawling.
+ */
+public class DynamicCrawlerHelper {
```
Comment (rzo1) [inline crawler4j-core/src/main/java/edu/uci/ics/crawler4j/crawler/dynamic/DynamicCrawlerHelper.java:20]: What do you think about introducing a new module `crawler4j-extensions` which contains `crawler4j-extensions-dynamic` (not sure about the names) to encapsulate everything related to dynamic scraping? That would make it possible to be lightweight or getting the full dynamic crawling capacity? wdyt?

File: crawler4j-core/src/main/java/edu/uci/ics/crawler4j/crawler/dynamic/DynamicWebCrawler.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
+package edu.uci.ics.crawler4j.crawler.dynamic;
```
Comment (rzo1) [inline crawler4j-core/src/main/java/edu/uci/ics/crawler4j/crawler/dynamic/DynamicWebCrawler.java:1]: Missing Header

File: crawler4j-core/src/main/java/edu/uci/ics/crawler4j/crawler/dynamic/DynamicWebCrawler.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    public void onBeforeExit() {
+        this.webDriver.close();
+    }
+
+    /**
+     * Default options for the headless browser
+     * @param config the {@link DynamicCrawlConfig} where the user agent is defined.
+     * @return the default options for the headless browser
+     */
+    public List<String> defaultHeadlessBrowserOptions(DynamicCrawlConfig config) {
```
Comment (rzo1) [inline crawler4j-core/src/main/java/edu/uci/ics/crawler4j/crawler/dynamic/DynamicWebCrawler.java:41]: Does this need to be `public` ?

File: crawler4j-core/src/main/java/edu/uci/ics/crawler4j/crawler/dynamic/DynamicWebCrawler.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        return Arrays.asList(
```
Comment (rzo1) [inline crawler4j-core/src/main/java/edu/uci/ics/crawler4j/crawler/dynamic/DynamicWebCrawler.java:42]: What happens if an options is specified multiple times, i.e. a user puts in `--window-size=800,600`. Which option has preference? Should we check for duplicates and replace?

File: crawler4j-core/src/main/java/edu/uci/ics/crawler4j/crawler/dynamic/DynamicWebCrawler.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+                "--user-agent=" + config.getWebDriverUserAgent());
+    }
+
+    /**
+     * Browser options defined in {@link DynamicCrawlConfig#getWebDriverOptions()} are concatenated to default options.
+     */
+    private String[] headlessBrowserOptions(DynamicCrawlConfig config) {
+        List<String> options = new ArrayList<>();
+        options.addAll(defaultHeadlessBrowserOptions(config));
+        options.addAll(config.getWebDriverOptions());
```
Comment (rzo1) [inline crawler4j-core/src/main/java/edu/uci/ics/crawler4j/crawler/dynamic/DynamicWebCrawler.java:56]: see above

File: crawler4j-core/src/main/java/edu/uci/ics/crawler4j/crawler/dynamic/DynamicWebCrawler.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+        return options.toArray(String[]::new);
+    }
+
+    private WebDriver newWebDriverInstance(DynamicCrawlConfig config) {
+        System.setProperty("webdriver.firefox.driver", config.getWebDriverPath());
```
Comment (rzo1) [inline crawler4j-core/src/main/java/edu/uci/ics/crawler4j/crawler/dynamic/DynamicWebCrawler.java:61]: Is this required if a ChromeDriver is used?

File: crawler4j-core/src/main/java/edu/uci/ics/crawler4j/parser/dynamic/DynamicParser.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
+package edu.uci.ics.crawler4j.parser.dynamic;
```
Comment (rzo1) [inline crawler4j-core/src/main/java/edu/uci/ics/crawler4j/parser/dynamic/DynamicParser.java:1]: Missing header

File: crawler4j-core/src/main/java/edu/uci/ics/crawler4j/parser/dynamic/DynamicTikaHtmlParser.java [hunk 13]
```diff
@@ -0,0 +1,95 @@
+package edu.uci.ics.crawler4j.parser.dynamic;
```
Comment (rzo1) [inline crawler4j-core/src/main/java/edu/uci/ics/crawler4j/parser/dynamic/DynamicTikaHtmlParser.java:1]: Missing header

**Review by rzo1 — COMMENTED** [review]
Thanks for this contributions. I didn't had time to test it (yet) but will obviously run a little crawl with it :-)

I have added some comments / questions.

**Comment by rzo1** [comment]
+ maybe we should add some sort of (basic) tests for it.

File: crawler4j-commons/src/main/java/edu/uci/ics/crawler4j/crawler/DynamicCrawlConfig.java (hunk 2 above)
Comment (ngsoftwaredev) [reply crawler4j-commons/src/main/java/edu/uci/ics/crawler4j/crawler/DynamicCrawlConfig.java:27]: Right, fixed.

File: crawler4j-commons/src/main/java/edu/uci/ics/crawler4j/crawler/DynamicCrawlConfig.java (hunk 4 above)
Comment (ngsoftwaredev) [reply crawler4j-commons/src/main/java/edu/uci/ics/crawler4j/crawler/DynamicCrawlConfig.java:54]: Hmmm actually no, you're right. I'll remove it.

File: crawler4j-core/src/main/java/edu/uci/ics/crawler4j/crawler/dynamic/DynamicWebCrawler.java (hunk 11 above)
Comment (ngsoftwaredev) [reply crawler4j-core/src/main/java/edu/uci/ics/crawler4j/crawler/dynamic/DynamicWebCrawler.java:80]: That's a silly bug, the property differs with the driver used, I'll fix it.

File: crawler4j-core/src/main/java/edu/uci/ics/crawler4j/crawler/dynamic/DynamicWebCrawler.java (hunk 9 above)
Comment (ngsoftwaredev) [reply crawler4j-core/src/main/java/edu/uci/ics/crawler4j/crawler/dynamic/DynamicWebCrawler.java]: That's not great indeed. I've replaced the `List` with a `Map<String, Optional<String>> ` and now options defined in the crawl config will override default ones.

File: crawler4j-core/src/main/java/edu/uci/ics/crawler4j/crawler/dynamic/DynamicWebCrawler.java (hunk 10 above)
Comment (ngsoftwaredev) [reply crawler4j-core/src/main/java/edu/uci/ics/crawler4j/crawler/dynamic/DynamicWebCrawler.java]: Fixed, see above too.

**Comment by ngsoftwaredev** [comment]
OK so I've fixed the defects you found. I'll see for adding some basic tests. Not quite sure what I can unit test though.

**Comment by rzo1** [comment]
> I'll see for adding some basic tests. Not quite sure what I can unit test though.

I was thinking more into integration testing, but might be a bit hard to implement for the gain we get from it; something like spin up a webserver  (testcontainers maybe) + some silly JS-based webpage. But don't think it has any priority ;-)

File: crawler4j-examples/crawler4j-examples-dynamic/pom.xml [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+        <groupId>de.hs-heilbronn.mi</groupId>
+        <version>5.0.3-SNAPSHOT</version>
+    </parent>
+    <artifactId>crawler4j-examples-dynamic</artifactId>
+    <name>${project.groupId}:${project.artifactId}</name>
+
+    <modelVersion>4.0.0</modelVersion>
+
+    <dependencies>
+        <dependency>
```
Comment (rzo1) [inline crawler4j-examples/crawler4j-examples-dynamic/pom.xml:14]: Actually we only need one of them but it is ok for the example

**Comment by rzo1** [comment]
> OK so I've fixed the defects you found. Let me know if the changes make sense to you.

The changes look good to me. I will try to find some time to run some on our JS-based university website.

**Comment by rzo1** [comment]
It is a flaky test ;-) - from a unit test side everything is ok.

**Comment by ngsoftwaredev** [comment]
OK, I realized I should make the wait for dynamic content loading more generic, I'll make a last commit to address this.

**Comment by rzo1** [comment]
I downloaded https://github.com/mozilla/geckodriver/releases/download/v0.33.0/geckodriver-v0.33.0-linux64.tar.gz and tried to run the given example.
It resulted in a 

```bash

2023-08-24 07:36:14,785 INFO  [main] frontier.SleepycatFrontierConfiguration (SleepycatFrontierConfiguration.java:66) - Deleted contents of: /tmp/crawler4j/frontier ( as you have configured resumable crawling to false )
2023-08-24 07:36:15,940 INFO  [main] crawler.CrawlController (CrawlController.java:248) - Crawler 1 started
2023-08-24 07:36:15,941 INFO  [main] crawler.CrawlController (CrawlController.java:248) - Crawler 2 started
Exception in thread "main" java.lang.RuntimeException: error on thread [Crawler 1]
	at edu.uci.ics.crawler4j.crawler.CrawlController.lambda$start$0(CrawlController.java:280)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.openqa.selenium.SessionNotCreatedException: Could not start a new session. Response code 400. Message: binary is not a Firefox executable 
Host info: host: 'node-147', ip: '127.0.1.1'
Build info: version: '4.11.0', revision: '040bc5406b'
System info: os.name: 'Linux', os.arch: 'amd64', os.version: '6.2.0-26-generic', java.version: '11.0.20'
Driver info: org.openqa.selenium.firefox.FirefoxDriver
Command: [null, newSession {capabilities=[Capabilities {acceptInsecureCerts: true, browserName: firefox, moz:debuggerAddress: true, moz:firefoxOptions: {args: [--disable-gpu, --user-agent=Mozilla/5.0 (X..., --headless, --window-size=1920,1080, --ignore-certificate-errors], binary: /usr/bin/firefox}}]}]
	at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:140)
	at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:96)
	at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:68)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:163)
	at org.openqa.selenium.remote.service.DriverCommandExecutor.invokeExecute(DriverCommandExecutor.java:196)
	at org.openqa.selenium.remote.service.DriverCommandExecutor.execute(DriverCommandExecutor.java:171)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:518)
	at org.openqa.selenium.remote.RemoteWebDriver.startSession(RemoteWebDriver.java:232)
	at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:159)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:156)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:151)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:132)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:127)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:112)
	at edu.uci.ics.crawler4j.crawler.dynamic.DynamicWebCrawler.newWebDriverInstance(DynamicWebCrawler.java:65)
	at edu.uci.ics.crawler4j.crawler.dynamic.DynamicWebCrawler.onStart(DynamicWebCrawler.java:44)
	at edu.uci.ics.crawler4j.crawler.WebCrawler.run(WebCrawler.java:294)
	... 1 more

Process finished with exit code 1

```

It seems, that the default Ubuntu Firefox installation is coming via `snap`. I switched that to the official Mozilla ppa but no luck either.

```
$ whereis firefox
firefox: /usr/bin/firefox /usr/lib/firefox /etc/firefox /snap/bin/firefox /usr/share/man/man1/firefox.1.gz
```

After this switch, I am getting

```bash
Exception in thread "main" java.lang.RuntimeException: error on thread [Crawler 1]
	at edu.uci.ics.crawler4j.crawler.CrawlController.lambda$start$0(CrawlController.java:280)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.openqa.selenium.SessionNotCreatedException: Could not start a new session. Response code 400. Message: binary is not a Firefox executable 
Host info: host: 'node-147', ip: '127.0.1.1'
Build info: version: '4.11.0', revision: '040bc5406b'
System info: os.name: 'Linux', os.arch: 'amd64', os.version: '6.2.0-26-generic', java.version: '11.0.20'
Driver info: org.openqa.selenium.firefox.FirefoxDriver
Command: [null, newSession {capabilities=[Capabilities {acceptInsecureCerts: true, browserName: firefox, moz:debuggerAddress: true, moz:firefoxOptions: {args: [--user-agent=Mozilla/5.0 (X..., --headless, --window-size=1920,1080, --ignore-certificate-errors, --disable-gpu], binary: /usr/lib/firefox/firefox.sh}}]}]
	at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:140)
	at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:96)
	at org.openqa.selenium.remote.ProtocolHandshake.createSession(ProtocolHandshake.java:68)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:163)
	at org.openqa.selenium.remote.service.DriverCommandExecutor.invokeExecute(DriverCommandExecutor.java:196)
	at org.openqa.selenium.remote.service.DriverCommandExecutor.execute(DriverCommandExecutor.java:171)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:518)
	at org.openqa.selenium.remote.RemoteWebDriver.startSession(RemoteWebDriver.java:232)
	at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:159)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:156)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:151)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:132)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:127)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:112)
	at edu.uci.ics.crawler4j.crawler.dynamic.DynamicWebCrawler.newWebDriverInstance(DynamicWebCrawler.java:66)
	at edu.uci.ics.crawler4j.crawler.dynamic.DynamicWebCrawler.onStart(DynamicWebCrawler.java:45)
	at edu.uci.ics.crawler4j.crawler.WebCrawler.run(WebCrawler.java:294)
	... 1 more
```

This looks interesting because `/usr/bin/firefox` is just a symlink to the sh script in a stock.

It seems, that I am doing something obviously wrong. Any pointers?

**Comment by rzo1** [comment]
Ok seems to be Ubuntu 22 - snap - yada yada related ;-) - will test in a VM setup.

**Comment by rzo1** [comment]
Looks like I need to setup an additional linux-based VM. Any distro suggestions for testing? ;-)

**Comment by ngsoftwaredev** [comment]
Actually I did my testing with _chromedriver_, I have to look into Geckodriver a bit more.

What I have already seen is that there is a way to explicitly set the Firefox binary. That might help, I should probably expose it. I'll let you know how I fare, it might be a couple days though before I can resume work on this.

**Comment by rzo1** [comment]
No problem. I was able to get it to work using the chromedriver ;-) - think we would need to mention https://googlechromelabs.github.io/chrome-for-testing/ in the README.

**Comment by rzo1** [comment]
@schwzr If you want to give it a try to provide feedback, feel free ;-)

File: crawler4j-commons/src/main/java/edu/uci/ics/crawler4j/crawler/DynamicCrawlConfig.java [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+
+    /**
+     * If dynamic content crawling is used, the maximum time to wait for dynamic content loading in seconds.
+     */
+    private int maxWaitForDynamicContentInSeconds = 2;
+
+    /**
+     * Absolute path to the driver binary on the local filesystem.
+     */
+    private Path webDriverPath;
```
Comment (valfirst) [inline crawler4j-commons/src/main/java/edu/uci/ics/crawler4j/crawler/DynamicCrawlConfig.java:50]: AFAIK it's not needed nowadays: https://www.selenium.dev/documentation/selenium_manager/#automated-driver-management

**Comment by ngsoftwaredev** [comment]
Thanks for the heads up! It's been quite a while since I issued the PR, and
more work is to be done on it.

I've lacked the time so far but fingers crossed I hope I can get back to it
soon.

**Pull request closed without merging**